src/UhdmAst.cpp

```diff
@@ -247,6 +247,32 @@ AstNode* get_value_as_node(vpiHandle obj_h, bool need_decompile = false) {
     return valueNodep;
 }
 
+AstNodeDType* get_packed_array_dtype(vpiHandle obj_h, UhdmShared& shared, AstBasicDType* basic) {
+    AstNodeDType* dtype = nullptr;
+    AstRange* rangeNode = nullptr;
+    std::stack<AstRange*> range_stack;
+    visit_one_to_many({vpiRange}, obj_h, shared, [&](AstNode* node) {
+        rangeNode = reinterpret_cast<AstRange*>(node);
+        range_stack.push(rangeNode);
+    });
+    rangeNode = nullptr;
+    while (!range_stack.empty()) {
+        if (rangeNode == nullptr) {
+            // Last node serves for basic type
+            rangeNode = range_stack.top();
+            basic->rangep(rangeNode);
+            dtype = basic;
+        } else {
+            // For other levels create a PackedArray
+            rangeNode = range_stack.top();
+            dtype = new AstPackArrayDType(new FileLine("uhdm"), VFlagChildDType(), dtype,
+                                          rangeNode);
+        }
+        range_stack.pop();
+    }
+    return dtype;
```

Comment:
This can fail silently and return a `nullptr` if the node does not have a `vpiRange` (and that can happen for types other than `vpiPackedArrayTypespec`).

Reply:
The only place when that function is called and type is not `vpiPackedArrayTypespec` is https://github.com/antmicro/verilator/pull/262/files#diff-eab5a9ad66f2411094af89dec2064f768624fd994a7a51ea6244c67cfc794c15R399
There is test if the returned type is `nullptr` and if it is, then `basic` type taken as dtype.